### ISSUE TITLE
Some Questions on ConvLab-Module-NLG-SCLSTM

### ISSUE THREAD

**Issue by ToSev7en** [ConvLab/ConvLab#36]
Hi!  Thank you for sharing such a good toolkit! 
Recently, when I looked into the NLG module, I found some confusing parts:

1. There are several evaluate code in ConvLab NLG module, including:

- **the `evaluate.py` in the `ConvLab/convlab/modules/nlg/multiwoz/evaluate.py`**

<img width="997" alt="image" src="https://user-images.githubusercontent.com/9492778/62270690-31b4a980-b469-11e9-963a-d631dc1d82ab.png">

- **the `bleu.py` in the `ConvLab/convlab/modules/nlg/multiwoz/sc_lstm/bleu.py`**
<img width="904" alt="image" src="https://user-images.githubusercontent.com/9492778/62271064-3c237300-b46a-11e9-883a-576efe378a1b.png">

I don't know which code is the right evaluation code. They all have different methods for calculating BLEU Score. 

And in the `evaluate.py` , line 52-57, I think the way to construct `refs` may have some problems, it make the Corpus Level BLEU score on `test.json` higher. I got the Corpus Level BLEU score `0.49` while in another paper the Corpus Level BLEU score is 0.18.

<img width="1188" alt="image" src="https://user-images.githubusercontent.com/9492778/62271412-219dc980-b46b-11e9-9c6b-bf6ecbd3fbb4.png">

Looking forward to your reply. Thanks!

**Comment by zqwerty** [comment]
Thanks for the question!

In general, you should use `ConvLab/convlab/modules/nlg/multiwoz/evaluate.py`. 

The way we calculate bleu is different from machine translation task. We group the sentence by their dialog act. For example, if `Inform-Hotel-Addr` dialog act has 3 golden sentences [r1,r2,r3], then for each generated sentence, its reference sentences is [r1,r2,r3].

The differences between `ConvLab/convlab/modules/nlg/multiwoz/evaluate.py` and `ConvLab/convlab/modules/nlg/multiwoz/sc_lstm/bleu.py`:
- The former replaces `value` in the sentence with corresponding `dialog_act-slot`, while the latter uses the delexicalized output (see `sclstm.res`) of sclstm directly. So the former one can be used for other NLG models.
- The former generates one sentence each time(beam_size=1), which is slower compared with using batch.

**Comment by ToSev7en** [comment]
@zqwerty Thank you !  

It seems that I understand the way you  group the sentence by dialog act. By this way, a system response generated by one dialog act (cause the beam_size=1 as you said ) may have multiple refs.

Comparing to another way that calculates a system response generated by one dialog act with it's  only one golden system response, the way used in Convlab may get a higher BLEU4 score, right ?

And one more question: In `ConvLab/convlab/modules/nlg/multiwoz/evaluate.py`, the SCLSTM model is load from a remote source. 

```python
print("Loading", model_name)
if model_name == 'SCLSTM':
    model_sys = SCLSTM(model_file="https://convlab.blob.core.windows.net/models/nlg-sclstm-multiwoz.zip")
```

And in `config.cfg`, it shows that the model was only trained and evaluated on `Boo_ResDataSplitRand0925.json` rather than `train.json`. Would this situation be a problem?
```
[DATA]
vocab_file =	%(dir)s/resource/vocab.txt
feat_file =		%(dir)s/resource/feat.json
text_file =		%(dir)s/resource/text.json
template_file =	%(dir)s/resource/template.txt
dataSplit_file= %(dir)s/resource/Boo_ResDataSplitRand0925.json
batch_size = 256
shuffle = true
dir = 

[MODEL]
dec_type = sclstm
hidden_size = 100
dropout = 0.25
clip = 0.5
learning_rate = 0.001

[TRAINING]
model_epoch = best
n_epochs = 75
```

**Comment by zqwerty** [comment]
> Comparing to another way that calculates a system response generated by one dialog act with it's only one golden system response, the way used in Convlab may get a higher BLEU4 score, right ?

Yes.

**Comment by truthless11** [comment]
> And in config.cfg, it shows that the model was only trained and evaluated on Boo_ResDataSplitRand0925.json rather than train.json. Would this situation be a problem?

We migrate SCLSTM from [multiwoz benchmark](https://github.com/andy194673/nlg-sclstm-multiwoz) where 'Boo_ResDataSplitRand0925.json' is used and this json file randomly split the dataset into train/valid/test set, so we don't think it would cause a big problem.
